subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java

```diff
@@ -146,27 +147,35 @@ public ClassSetAnalysisData reduceToTypesAffecting(ClassSetAnalysisData other) {
             }
         }
 
+        Set<String> usedConstantSources = compilerApiData.isSupportsConstantsMapping()
+            ? compilerApiData.getConstantToClassMapping().getConstantDependents().keySet()
+            : classesToConstants.keySet();
+        
+        usedClasses.addAll(usedConstantSources);
+
         Map<String, HashCode> classHashes = new HashMap<>(usedClasses.size());
         Map<String, DependentsSet> dependents = new HashMap<>(usedClasses.size());
         Map<String, IntSet> classesToConstants = new HashMap<>(usedClasses.size());
         for (String usedClass : usedClasses) {
             HashCode hash = this.classHashes.get(usedClass);
             if (hash != null) {
                 classHashes.put(usedClass, hash);
-            }
-            DependentsSet dependentsSet = this.dependents.get(usedClass);
-            if (dependentsSet != null) {
-                if (dependentsSet.isDependencyToAll()) {
-                    dependents.put(usedClass, dependentsSet);
-                } else {
-                    Set<String> usedAccessibleClasses = new HashSet<>(dependentsSet.getAccessibleDependentClasses());
-                    usedAccessibleClasses.retainAll(usedClasses);
-                    dependents.put(usedClass, DependentsSet.dependentClasses(Collections.emptySet(), usedAccessibleClasses));
+                DependentsSet dependentsSet = this.dependents.get(usedClass);
```

Comment:
Moved this up into the if-statement, so we only keep classes that are actually on the classpath. This exludes classes that are referenced, but not on the classpath. This happens e.g. with the `gradleApi` JAR, which has references to types that it doesn't include.

Reply:
do we have a test case covering this change?

Reply:
It's a performance optimization, so the perf tests should handle regressions here.